### PR TITLE
Add banner to app

### PR DESCRIPTION
**Pull request by kcvan**
## Description
Why did you write this code?
add email and phone verification banner to app
## How to Test
- [ ] Create a new account
- [ ] Banner should show up on mobile/tablet/desktop
- [ ] Banner should be closable by clicking the `X` button
- [ ] Banner should lead you to account verification page
- [ ] Verifying your phone number and email, then refreshing the browser should remove the banner

Which devices did you test on?
- [x] Chrome on Mac
- [ ] Chrome on PC
- [ ] Firefox on Mac
- [ ] Firefox on PC
- [ ] Safari iPhone
- [ ] Chrome Android

## REVIEWERS:
Check against these principles:

### High level
Does this code need to be written?
What are the alternatives?
Will this implementation become a support issue?
How much error margin does this solution have?

### Code
* Does the code follow industry standards?
JS: https://github.com/airbnb/javascript
React: https://github.com/airbnb/javascript/tree/master/react
https://github.com/vasanthk/react-bits
Documentation headers: http://usejsdoc.org/index.html

* Is there duplicated code? Can it be refactored into a shared method?
* Is the code consistent with our project?
* Are there unit tests? Do they test the states?
* Is the person refactoring another developer's code? If possible, did the original developer approve?

### Variables/Naming:
* Would the variable type led to future edge cases?
* Are the variable naming clear? Would the value contain something other than what the name describes.

## Further Work
Will this have future work?
To enable live closing of the banner, we might need to alter the Firebase action routes to pull data from Firebase and close the banner based on the information.
## Learnings
Did you learn anything here you want to share with the team?
Since rebranding our application, we've been starting to use react hooks almost everywhere in our app.  This means that the way we originally used `Banner` and the `BannerContext` would no longer work in non-class components. 

Although using `useContext` and `useReducer` by themselves is pretty straightforward, combining the context state with userReducer proved to be difficult.  There aren't many working examples online, and even fewer examples _with_ typescript.

The best part is that the context api is backwards compatible, so we can still use the provider/consumer methodology on some parts of the app, and the react hooks way in others.

### CODE REVIEW

File: src/legacy/Legacy.routes.tsx [hunk 1]
```diff
@@ -1,52 +1,45 @@
-import * as React from 'react';
-import { Route, Switch } from 'react-router-dom';
+// import * as React from 'react';
```
Comment (woeltjen) [inline src/legacy/Legacy.routes.tsx]: This file looks all commented out; should it be deleted?

Comment (kcvan) [reply src/legacy/Legacy.routes.tsx]: I was thinking about deleting it in another pr. Wanted to make sure that I wasn't deleting anything that might need to be referenced later but we can always check the git tag.  Will delete here 👍

File: src/utils/workBannerUtility.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    to: '/account/verification',
+  }
+
+  if (userVerified) {
+    await bannerDispatch({ type: 'close' });
+  }
+
+  if (userUnverified) {
+    await bannerDispatch({ type: 'set', payload: options });
+    await bannerDispatch({ type: 'open' });
```
Comment (jerminatorhits) [inline src/utils/workBannerUtility.ts]: doesn't this happen synchronously without `await`

Comment (jerminatorhits) [reply src/utils/workBannerUtility.ts]: or maybe like `this.setState` it might batch the dispatch changes together

Comment (kcvan) [reply src/utils/workBannerUtility.ts]: That's a good question.  Originally we used the await in case this because a race condition on whether the banner would set banner props first or open first.  I think in the background `useReducer` still uses setState, so this might still hold true.  An option we can do is wrap the dispatch calls with `new Promise` but I think that might add unnecessary bulk.  We can try to remove it for now and see if the banner every loads before the message.  Thoughts?

**Comment by 0xTail** [comment]
👍

File: src/utils/workBannerUtility.ts (hunk 2 above)
Comment (jerminatorhits) [reply src/utils/workBannerUtility.ts]: ya sounds good to me 👍

File: src/HOCs/FirebaseProvider.tsx [hunk 3]
```diff
@@ -11,8 +11,8 @@
 
 import * as React from 'react';
 
-import { BannerConsumerProps } from 'HOCs/BannerProvider';
-import { showAccountVerificationBanner } from 'utils/bannerUtility';
+import { WorkBannerConsumerProps } from 'HOCs/WorkBannerProvider';
```
Comment (tc) [inline src/HOCs/FirebaseProvider.tsx]: Does this need to be prefixed `Work` if there is only one now?

Comment (kcvan) [reply src/HOCs/FirebaseProvider.tsx]: Good point! Will remove `Work` from the Banner

File: src/legacy/shared/Banner/Banner.tsx [hunk 4]
```diff
@@ -39,18 +37,18 @@ const Banner = (props: Props) => {
...
-          );
-        }}
-      </AppConsumer>
+      <CloseButton
+        className="d-flex d-md-none close"
+        height="32px"
+        iconColor="upper"
+        onClose={props.onClose}
+        width="32px" />
+      <CloseButton
```
Comment (tc) [inline src/legacy/shared/Banner/Banner.tsx]: Is this a copy pasta error? i see two close buttons

Comment (kcvan) [reply src/legacy/shared/Banner/Banner.tsx]: When I removed the `AppConsumer` we didn't have a way to check if the user was on `mobile`, `tablet`, or `desktop` so I used two components and used `display: none` depending on the screen size (given by bootstrap)

File: src/utils/workBannerUtility.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { WorkBannerConsumerProps } from 'HOCs/WorkBannerProvider';
+
+export const ACCOUNT_VERIFICATION = 'Click here to verify your email and phone number to start booking!';
+
+export const showAccountVerificationBanner = (
+  isUserVerified: boolean,
+  bannerDispatch: WorkBannerConsumerProps['bannerDispatch'],
+  bannerState: WorkBannerConsumerProps['bannerState']): void => {
+  const userVerified = isUserVerified && bannerState.show;
```
Comment (tc) [inline src/utils/workBannerUtility.ts]: this can confusing: `userVerified` is equal to `isUserVerified` AND a banner state.
I don't understand why userVerified is conditional on the banner state which is a UI variable.

Did you mean `userVerified` be named something else?

Comment (kcvan) [reply src/utils/workBannerUtility.ts]: That's a good point.  We're receiving the return of `hasCompletedVerification(user)` which will tell us if a user has finished phone and email verification.  so if the user is verified and the banner is shower we want to close the banner, and the opposite if the user is unverified.  I think I can fix up the names a bit to make it more clear 🙏

File: src/legacy/shared/Banner/Banner.tsx (hunk 4 above)
Comment (tc) [reply src/legacy/shared/Banner/Banner.tsx]: ah i see; can we assign the `32px` to a variable like `mobileButtonSize` to make it more apparent you are doing this?

Comment (kcvan) [reply src/legacy/shared/Banner/Banner.tsx]: Sounds good!

File: config/webpack.common.js [hunk 6]
```diff
@@ -15,7 +15,6 @@ const nodeModules = path.resolve(__dirname, '..', 'node_modules');
 module.exports = {
   entry: {
     index: `${root}/index.tsx`,
-    legacy: `${root}/legacy.tsx`,
```
Comment (tc) [inline config/webpack.common.js]: it would have been better if these legacy code deletion were in a separate PR since they are unrelated changes.

File: src/utils/bannerUtility.ts [hunk 7]
```diff
@@ -2,22 +2,23 @@ import { BannerConsumerProps } from 'HOCs/BannerProvider';
...
-  if (userVerified) {
-    await bannerActions.closeBanner();
+  if (userVerified && bannerState.show) {
+    bannerDispatch({ type: 'close' });
   }
 
-  if (userUnverified) {
-    await bannerActions.setBannerOptions(options);
-    await bannerActions.openBanner();
+  if (!(userVerified || bannerState.show)) {
```
Comment (tc) [inline src/utils/bannerUtility.ts]: Is this more understandable? `!userVerified && !bannerState.show`
show the banner "if the user is not verified and the banner is not showing"

File: config/webpack.common.js (hunk 6 above)
Comment (kcvan) [reply config/webpack.common.js]: Agreed.  Removing the original banner caused a lot of errors that eventually lead to be taking out a bunch of legacy files.  Commenting out the code seemed about the same so I decided to just remove the code.

File: src/utils/bannerUtility.ts (hunk 7 above)
Comment (kcvan) [reply src/utils/bannerUtility.ts]: Definitely is, fixing in next commit 🙏
